### PR TITLE
Use raw values, not display text, for Entity property values

### LINKED ISSUE #6510: Dates saved to offline Entity properties have the wrong format when used in XPath expressions
#### ODK Collect version
v2024.1.0 beta 8

#### Problem description
When an offline update is made to an Entity and this update represents a date, the date is converted to DD/MM/YY HH:MM format instead of remaining in ISO 8601 format. This means that any expressions trying to use that date crashes. This is a local-only issue: once the Entity is updated on the server and synced back to the device, everything works as expected.

#### Steps to reproduce the problem
1. Create a form that saves a date to an Entity property and then later uses that date such as [this one](https://docs.google.com/spreadsheets/d/1_VsrftvmLgvuldUFtz2C1uxuUVSrMKRK8pKdZBZvNHw/edit?gid=1068911091#gid=1068911091)
2. Make sure that there is the `offline` column on the entities sheet and that its value is set to `yes`
3. Fill out and finalize a submission
4. Open the form again and see the crash
5. Open the Entities browser and the property has been incorrectly re-formated
Screenshot_20241114_154559

#### Expected behavior
Entity properties should always be treated as opaque strings. They should be saved and retrieved exactly as entered into the form.

#### Other information 

Something along the way is detecting a date and reformatting it. I don't think it's sqlite because I'm pretty sure sqlite stores dates in iso8601 format.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #6510 

#### Why is this the best possible solution? Were any other approaches considered?
There is no other solution to discuss. If we want to use plain, unformatted strings, we need to uncast the value and read the string.
There is one issue that needs to be addressed in javarosa I belive: https://github.com/getodk/collect/pull/6514/files#r1844165858

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Regardless of the data type used to store entities, the values saved to them should be plain, unformatted text, identical to what is stored in XML submissions. I think testing dates here is sufficient; we don't need to test other data types. Even if this issue affects other data types, the PR should address all cases.

#### Do we need any specific form for testing your changes? If so, please attach one.
I used:
[entities_with_dates_registration.xlsx](https://github.com/user-attachments/files/17814115/entities_with_dates_registration.xlsx)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/EntityFormTest.kt [hunk 1]
```diff
@@ -23,6 +23,22 @@ class EntityFormTest {
...
+        testDependencies.server.addForm("entities-with-dates-registration.xml")
+        testDependencies.server.addForm("entities-with-dates-follow-up.xml")
+
+        rule.withMatchExactlyProject(testDependencies.server.url)
+            .startBlankForm("Entities With Dates Registration")
+            .swipeToEndScreen()
+            .clickFinalize()
+
+            .startBlankForm("Entities With Dates Follow Up")
+            // .assertText("2024-11-15")
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/EntityFormTest.kt:37]: There is a problem with saving dates as labels. I fixed it here: https://github.com/getodk/collect/pull/6514/files#diff-dcf4e4fc03ff048616e3d821f4dc1ff53ea090480626d6516d0ac054b35c1446R35 but the problem is that such an element is identified as `DateTime` data type not just `Date`. It seems to be a bug in javarosa so I'm going to file an issue there.

Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/EntityFormTest.kt:37]: The issue is here: https://github.com/getodk/javarosa/issues/807

**Review by seadowg — APPROVED** [review]
The solution looks good as far as I can tell, although I wasn't familiar with the `uncast` method. 

I think we could test this at the unit level for `EntityFormFinalizationProcessor` and `EntityFormParser` instead of needing a connected test though. @grzesiek2010 could you make that change in a follow-up? That way we're not blocking this for test changes.

**Comment by grzesiek2010** [comment]
> @grzesiek2010 could you make that change in a follow-up? That way we're not blocking this for test changes.

Ok I will do that.

**Comment by srujner** [comment]
Verified on device with Android 14

**Comment by dbemke** [comment]
Verified on device with Android 10

**Pull request merged**